File: frontend/find-unused-files.js
import fs from "fs";
import path from "path";
import { execSync } from "child_process";
import { fileURLToPath } from "url";
import { dirname } from "path";
import { parse } from "@babel/parser";
import traverse from "@babel/traverse";
import { transformSync } from "@babel/core";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Configuration
const PROJECT_ROOT = path.join(__dirname, "..");
const IGNORE_DIRS = [
  "node_modules",
  ".git",
  "dist",
  "build",
  ".next",
  ".vercel",
  ".github",
  "__tests__",
  "__mocks__",
  "coverage",
  "public",
  "scripts",
  "types",
  ".vscode",
  ".idea",
  "cypress",
  "mocks",
  "stories",
  "test",
  "test-utils",
];

const EXTENSIONS = [".ts", ".tsx", ".js", ".jsx"];
const ALIASES = {
  "@": path.join(PROJECT_ROOT, "src"),
  "~": path.join(PROJECT_ROOT, "src"),
};

// Cache for parsed files
const fileCache = new Map();

// Resolve path with aliases
function resolvePath(importPath, currentFile) {
  // Skip node modules and absolute paths
  if (importPath.startsWith(".") || importPath.startsWith("..")) {
    return path.resolve(path.dirname(currentFile), importPath);
  }

  // Handle path aliases
  const [alias] = Object.keys(ALIASES).filter((a) => importPath.startsWith(a));
  if (alias) {
    return path.join(ALIASES[alias], importPath.replace(alias, ""));
  }

  return importPath;
}

// Get all TypeScript/JavaScript files
function getAllSourceFiles(dir, fileList = []) {
  const files = fs.readdirSync(dir);
  const relativeDir = path.relative(PROJECT_ROOT, dir);

  // Skip ignored directories
  if (IGNORE_DIRS.some((ignoreDir) => relativeDir.includes(ignoreDir))) {
    return fileList;
  }

  files.forEach((file) => {
    const filePath = path.join(dir, file);
    const stat = fs.statSync(filePath);
    const relativePath = path.relative(PROJECT_ROOT, filePath);

    if (IGNORE_DIRS.some((ignoreDir) => relativePath.includes(ignoreDir))) {
      return;
    }

    if (stat.isDirectory()) {
      getAllSourceFiles(filePath, fileList);
    } else if (EXTENSIONS.some((ext) => file.endsWith(ext))) {
      fileList.push(relativePath);
    }
  });

  return fileList;
}

// Get all import statements from a file using Babel
function getImports(filePath) {
  const fullPath = path.join(PROJECT_ROOT, filePath);

  // Check cache first
  if (fileCache.has(fullPath)) {
    return fileCache.get(fullPath);
  }

  try {
    const content = fs.readFileSync(fullPath, "utf-8");

    // Skip empty files or files that can't be parsed
    if (!content.trim()) {
      fileCache.set(fullPath, []);
      return [];
    }

    const ast = parse(content, {
      sourceType: "module",
      plugins: ["jsx", "typescript", "decorators-legacy"],
    });

    const imports = new Set();
    const dynamicImports = new Set();

    // Handle static imports and requires
    traverse(ast, {
      ImportDeclaration(path) {
        imports.add(path.node.source.value);
      },
      CallExpression(path) {
        if (path.node.callee.type === "Import") {
          if (path.node.arguments[0]?.type === "StringLiteral") {
            dynamicImports.add(path.node.arguments[0].value);
          }
        } else if (
          path.node.callee.type === "Identifier" &&
          path.node.callee.name === "require" &&
          path.node.arguments[0]?.type === "StringLiteral"
        ) {
          imports.add(path.node.arguments[0].value);
        }
      },
    });

    // Handle dynamic imports in JSX
    const jsxContent =
      transformSync(content, {
        plugins: ["@babel/plugin-syntax-jsx"],
        presets: ["@babel/preset-react"],
      })?.code || "";

    const dynamicImportRegex = /import\(['"]([^'"]+)['"]\)/g;
    let match;
    while ((match = dynamicImportRegex.exec(jsxContent)) !== null) {
      dynamicImports.add(match[1]);
    }

    const allImports = [...imports, ...dynamicImports]
      .filter((imp) => !imp.startsWith("http"))
      .filter((imp) => !imp.startsWith("data:"))
      .filter((imp) => !imp.startsWith("file:"))
      .map((imp) => resolvePath(imp, filePath));

    fileCache.set(fullPath, allImports);
    return allImports;
  } catch (error) {
    console.error(`Error parsing ${filePath}:`, error.message);
    fileCache.set(fullPath, []);
    return [];
  }
}

// Main function
async function findUnusedFiles() {
  console.log("🔍 Scanning for unused files...\n");

  try {
    // Install required dependencies if not present
    try {
      require.resolve("@babel/parser");
    } catch (e) {
      console.log("Installing required dependencies...");
      execSync(
        "npm install --save-dev @babel/parser @babel/traverse @babel/core @babel/preset-react @babel/plugin-syntax-jsx",
        {
          cwd: PROJECT_ROOT,
          stdio: "inherit",
        },
      );
    }

    const { parse } = await import("@babel/parser");
    const traverse = (await import("@babel/traverse")).default;
    const { transformSync } = await import("@babel/core");

    const allFiles = [];
    getAllSourceFiles(PROJECT_ROOT, allFiles);

    const fileUsage = new Map();
    const fileDependencies = new Map();

    // Initialize all files
    allFiles.forEach((file) => {
      fileUsage.set(file, {
        used: false,
        usedBy: [],
        isEntry: [
          "src/main.tsx",
          "src/App.tsx",
          "src/index.tsx",
          "src/AppWrapper.tsx",
        ].includes(file),
      });
    });

    // First pass: build dependency graph
    for (const file of allFiles) {
      const imports = getImports(file);
      fileDependencies.set(file, imports);
    }

    // Second pass: resolve dependencies
    function markAsUsed(file, usedBy = null) {
      if (!fileUsage.has(file)) return;

      const fileInfo = fileUsage.get(file);
      if (fileInfo.used && !usedBy) return; // Already processed

      fileInfo.used = true;
      if (usedBy) {
        fileInfo.usedBy.push(usedBy);
      }

      // Mark all files this file imports as used
      const imports = fileDependencies.get(file) || [];
      imports.forEach((imp) => {
        // Try to find the actual file path
        const possibleFiles = [
          imp,
          `${imp}.ts`,
          `${imp}.tsx`,
          `${imp}.js`,
          `${imp}.jsx`,
          path.join(imp, "index.ts"),
          path.join(imp, "index.tsx"),
          path.join(imp, "index.js"),
          path.join(imp, "index.jsx"),
        ];

        for (const possibleFile of possibleFiles) {
          if (fileUsage.has(possibleFile)) {
            markAsUsed(possibleFile, file);
            break;
          }
        }
      });
    }

    // Start with entry points
    for (const [file, info] of fileUsage) {
      if (info.isEntry) {
        markAsUsed(file);
      }
    }

    // Mark all files in public directory as used
    allFiles.forEach((file) => {
      if (file.startsWith("public/") || file.includes("/public/")) {
        fileUsage.get(file).used = true;
      }
    });

    // Get unused files
    const unusedFiles = [];
    const usedFiles = [];

    fileUsage.forEach((value, key) => {
      if (value.used) {
        usedFiles.push(key);
      } else {
        unusedFiles.push(key);
      }
    });

    console.log(`✅ Used files (${usedFiles.length}):`);
    usedFiles.sort().forEach((file) => {
      console.log(`- ${file}`);
    });

    console.log("\n🚫 Potentially unused files:");
    unusedFiles.sort().forEach((file) => {
      console.log(`- ${file}`);
    });

    console.log(
      `\n📊 Found ${unusedFiles.length} potentially unused files out of ${allFiles.length} total files.`,
    );

    // Save results to a file
    fs.writeFileSync(
      path.join(PROJECT_ROOT, "unused-files-report.json"),
      JSON.stringify(
        {
          timestamp: new Date().toISOString(),
          totalFiles: allFiles.length,
          usedFiles: usedFiles.length,
          unusedFiles: unusedFiles.length,
          unusedFilesList: unusedFiles,
          fileUsage: Object.fromEntries(fileUsage),
        },
        null,
        2,
      ),
    );

    console.log("\n📝 Report saved to unused-files-report.json");
  } catch (error) {
    console.error("Error analyzing files:", error);
    process.exit(1);
  }
}

// Run the script
findUnusedFiles();

File: find_unused_files.py
import json
import os
import re
from pathlib import Path
from typing import Dict, List, Set


class UnusedFileFinder:
    def __init__(self, project_root: str):
        self.project_root = Path(project_root)
        self.src_dir = self.project_root / "frontend" / "src"
        self.ignored_dirs = {
            "node_modules",
            ".git",
            ".next",
            "dist",
            "build",
            "__pycache__",
            ".idea",
            ".vscode",
            "venv",
            ".venv",
            ".gitignore",
            ".eslintrc",
            ".prettierrc",
            ".env*",
            "*.log",
            "*.tmp",
            "*.bak",
            "*.swp",
            "*.swo",
        }
        self.used_files: Set[Path] = set()
        self.all_files: Set[Path] = set()
        self.import_pattern = re.compile(
            r"from\s+['\"]([^'\"]+)['\"]|import\s+.*\s+from\s+['\"]([^'\"]+)['\"]"
        )

    def is_ignored(self, path: Path) -> bool:
        """Check if a path should be ignored."""
        for part in path.parts:
            if part.startswith(".") and part not in [".", ".."]:
                return True
            if part in self.ignored_dirs:
                return True
            if part.endswith((".log", ".tmp", ".bak", ".swp", ".swo")):
                return True
        return False

    def find_all_files(self) -> None:
        """Find all relevant files in the project."""
        print("Scanning project files...")
        for root, _, files in os.walk(self.project_root):
            for file in files:
                file_path = Path(root) / file
                rel_path = file_path.relative_to(self.project_root)

                if not self.is_ignored(file_path) and file_path.suffix in {
                    ".ts",
                    ".tsx",
                    ".js",
                    ".jsx",
                }:
                    self.all_files.add(rel_path)

    def analyze_imports(self) -> None:
        """Analyze imports in all TypeScript/JavaScript files."""
        print("Analyzing imports...")
        for file_path in list(self.all_files):
            try:
                content = (self.project_root / file_path).read_text(encoding="utf-8")
                matches = self.import_pattern.finditer(content)

                for match in matches:
                    # Get either the first or second group that matched
                    import_path = match.group(1) or match.group(2)
                    if not import_path:
                        continue

                    # Convert import path to actual file path
                    if import_path.startswith("@/"):
                        import_path = import_path[2:]
                    elif import_path.startswith("."):
                        import_path = str(file_path.parent / import_path)

                    # Try to find the actual file
                    self._resolve_import(import_path, file_path.parent)

            except Exception as e:
                print(f"Error analyzing {file_path}: {e}")

    def _resolve_import(self, import_path: str, base_dir: Path) -> None:
        """Resolve an import path to an actual file."""
        # Try with .tsx, .ts, .js, .jsx extensions
        for ext in ["", ".tsx", ".ts", ".js", ".jsx"]:
            full_path = (self.project_root / base_dir / f"{import_path}{ext}").resolve()
            if full_path.exists():
                rel_path = full_path.relative_to(self.project_root)
                self.used_files.add(rel_path)
                return

    def find_unused_files(self) -> List[Dict]:
        """Find files that aren't imported anywhere."""
        unused = []
        for file_path in self.all_files:
            if file_path not in self.used_files and "test" not in str(file_path):
                unused.append(
                    {
                        "path": str(file_path),
                        "size_kb": (self.project_root / file_path).stat().st_size
                        / 1024,
                    }
                )
        return sorted(unused, key=lambda x: x["size_kb"], reverse=True)


def main():
    project_root = r"c:\Users\OMEN\OneDrive\Documents\vybe 2.0"
    output_file = "unused_files_analysis.md"

    finder = UnusedFileFinder(project_root)
    finder.find_all_files()
    finder.analyze_imports()
    unused_files = finder.find_unused_files()

    # Generate report
    total_size = sum(f["size_kb"] for f in unused_files)

    report = [
        "# Unused Files Analysis\n",
        "## Summary\n",
        f"- Total files analyzed: {len(finder.all_files)}",
        f"- Potentially unused files: {len(unused_files)}",
        f"- Estimated disk space to free: {total_size:.2f} KB\n",
        "## Potentially Unused Files\n",
    ]

    for file in unused_files:
        report.append(f"- `{file['path']}` ({file['size_kb']:.2f} KB)")

    with open(output_file, "w", encoding="utf-8") as f:
        f.write("\n".join(report))

    print(f"\n=== Analysis Complete ===")
    print(f"Found {len(finder.all_files)} files in the project")
    print(f"Identified {len(unused_files)} potentially unused files")
    print(f"See '{output_file}' for details")


if __name__ == "__main__":
    main()
